### PR TITLE
CSRF makes a comeback

### PR DESCRIPTION
**Pull request by michaelrtm**
## CSRF Makes a Comeback ##
This PR addresses #169 

I have written or updated a large amount of tests to cover front-end testing of all forms. I hope I haven't overlooked anything. All tests are passing, except for one test marked as incomplete as I haven't found a solution to testing reCaptcha.

 * To test CSRF effectively, one has to first enable CSRF testing on Tests. This can be done by commenting line 61 of `Illuminate\Foundation\Http\Middleware\VerifyCsrfToken` before testing.

---

`PublicProfileTest@user_can_be_contacted_from_profile()`

I cannot get this test to pass working with reCaptcha. If I remove the Captcha stuff from the controller, the test passes.

---
#### Forms / Tests ####
| Form                            | Test Coverage                                                |  |
|---------------------------------|--------------------------------------------------------------|-----|
| auth/passwords/reset            | AccountTest@password_reset_emails_are_sent_for_valid_users() | * |
| auth/passwords/email            | AccountTest@user_can_reset_their_password_from_email_link() | * |
| account/edit                    | AccountTest@user_can_update_their_profile() | * |
|                                 | AccountTest@user_can_update_their_profile_picture() | * |
| account/confirm-delete          | AccountTest@users_can_delete_their_accounts() | |
| account/public-profile/email    | PublicProfileTest@user_can_be_contacted_from_profile()  | ! |
| bios/create                     | Covered in BioTest | * |
| bios/edit                       | Covered in BioTest | * |
| conferences/edit                | ConferenceTest@user_can_create_conference() | ^ | 
| conferences/create              | ConferenceTest@user_can_edit_conference() | ^ |
| oauth/authorization-form        | Api/OAuthTest | | 
| partials/log-in-form            | AccountTest/users_can_log_in() | |
| partials/sign-up-form           | AccountTest/users_can_sign_up() | |
| submissions post only           | SubmissionTest@user_can_submit_talks_via_http |  ^ |
| talks/create                    | TalkTest/user_can_create_a_talk() | ^ | 
| talks/edit                      | TalkTest@user_can_save_a_new_revision_of_a_talk() | ^ |

*Any row marked with an asterisk is new*
*Any row marked with a caret is updated*

### CODE REVIEW

File: tests/BiosTest.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        $this->actingAs($user)
+            ->visit('/bios/' . $bio->id . '/edit')
+            ->type('Fresh Prince', '#nickname')
+            ->type('Born and raised in West Philidelphia, I spend a large majority of my time on the playground.', '#body')
+            ->select('yes', '#public')
+            ->press('Update');
+
+        $this->seeInDatabase('bios', [
+            'nickname' => 'Fresh Prince',
+            'body' => 'Born and raised in West Philidelphia, I spend a large majority of my time on the playground.'
```
Comment (imjohnbon) [inline tests/BiosTest.php]: Trailing comma should be added here.

File: tests/BiosTest.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /** @test */
+    function user_cannot_edit_a_bio_that_they_do_not_own()
+    {
+        $userA = Factory::create('user');
+        $userB = Factory::create('user');
+
+        $bio = Factory::build('bio', [
+            'user_id' => $userA->id
```
Comment (imjohnbon) [inline tests/BiosTest.php]: Trailing comma should be added here.

File: tests/BiosTest.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    /** @test */
+    function user_can_delete_their_bio()
+    {
+        $user = Factory::create('user');
+        $bio = App\Bio::create([
+            'user_id' => $user->id,
+            'nickname' => 'Jimmy Buffet',
+            'body' => '5 oclock somewhere',
+            'public' => 0
```
Comment (imjohnbon) [inline tests/BiosTest.php]: Trailing comma should be added here.

File: tests/BiosTest.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        ]);
+
+        $this->actingAs($user)->visit('/bios/' . $bio->id . '/delete');
+        $this->missingFromDatabase('bios', [
+            'nickname' => 'Jimmy Buffet',
+            'body' => '5 oclock somewhere'
```
Comment (imjohnbon) [inline tests/BiosTest.php]: Trailing comma should be added here.

File: tests/BiosTest.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    /** @test */
+    function user_cannot_delete_a_bio_they_dont_own()
+    {
+        $userA = Factory::create('user');
+        $userB = Factory::create('user');
+        $bio = App\Bio::create([
+            'user_id' => $userA->id,
+            'nickname' => 'Jimmy Buffet',
+            'body' => '5 oclock somewhere',
+            'public' => 0
```
Comment (imjohnbon) [inline tests/BiosTest.php]: Trailing comma should be added here.

File: tests/BiosTest.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        ]);
+
+        $this->actingAs($userB)->call('delete', '/bios/' . $bio->id);
+        $this->seeInDatabase('bios', [
+            'nickname' => 'Jimmy Buffet',
+            'body' => '5 oclock somewhere'
```
Comment (imjohnbon) [inline tests/BiosTest.php]: Trailing comma should be added here.

File: tests/SubmissionTest.php [hunk 7]
```diff
@@ -162,6 +163,7 @@ function user_can_submit_talks_via_http()
         $this->post('submissions', [
             'conferenceId' => $conference->id,
             'talkId' => $talk->id,
+            '_token' => csrf_token()
```
Comment (imjohnbon) [inline tests/SubmissionTest.php]: Trailing comma should be added here.

File: tests/TalkTest.php [hunk 8]
```diff
@@ -71,21 +71,27 @@ function user_talks_json_encode_without_keys()
...
+            ->press('Create');
+
+        $this->seeInDatabase('talk_revisions', [
             'title' => 'Your Best Talk Now',
             'type' => 'keynote',
             'level' => 'intermediate',
             'description' => 'No, really.',
             'length' => '123',
             'slides' => 'http://www.google.com/slides',
             'organizer_notes' => "It'll be awesome!"
```
Comment (imjohnbon) [inline tests/TalkTest.php]: Trailing comma should be added here.

File: tests/AccountTest.php [hunk 9]
```diff
@@ -52,21 +52,89 @@ function users_can_log_in()
...
+            'profile_intro' => 'It has been so long since I was in an X-Men movie',
+        ]);     
+    }
+
+    /** @test */
+    function user_can_update_their_profile_picture()
+    {
+        $image = __DIR__.'/stubs/test.jpg';
+        $user = Factory::create('user', [
+            'name' => 'Kevin Smith'
```
Comment (imjohnbon) [inline tests/AccountTest.php]: Trailing comma should be added here.

**Comment by imjohnbon** [comment]
@michaelrtm Hey! This looks great! Just left some notes, mostly minor formatting issues as well as 2 tests that need a slight adjustment to pass. Let me know if you have any questions!

File: tests/BiosTest.php [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            ->type('Some Nickname', '#nickname')
+            ->type('A big chunk of bio-friendly text', '#body')
+            ->select('no', '#public')
+            ->press('Create');
+            //->seePageIs('bios'); //not sure how to test the string /bios/x
+
+        $this->seeInDatabase('bios', [
+            'nickname' => 'Some Nickname',
+            'body' => 'A big chunk of bio-friendly text',
+            'public' => 'no'
```
Comment (michaelrtm) [inline tests/BiosTest.php]: This passes for me as 'no' but fails as 0, '0', false, 'false', 1, 'yes', 'random'.
Not sure where that may be originating from.

I can assert manually that creating that results in a 0 in the database, but for some reason only 'no' will pass.

File: tests/BiosTest.php [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            ->type('Some Nickname', '#nickname')
+            ->type('A big chunk of bio-friendly text', '#body')
+            ->select('yes', '#public')
+            ->press('Create');
+            //->seePageIs('bios'); //not sure how to test the string /bios/x
+
+        $this->seeInDatabase('bios', [
+            'nickname' => 'Some Nickname',
+            'body' => 'A big chunk of bio-friendly text',
+            'public' => 'yes'
```
Comment (michaelrtm) [inline tests/BiosTest.php]: As above, but obviously opposite - any ideas?

**Comment by michaelrtm** [comment]
I've left a few notes on the tests that are failing - they are passing for me! Might there be something wrong with my testing environment?

**Comment by mattstauffer** [comment]
@michaelrtm I believe it's because of a change that was made in master since you last pulled; when @imjohnbon runs his tests locally, he's running it in a version that has master merged in. This change converts it from a string to a boolean:

https://github.com/tightenco/symposium/commit/d7a581063ca7c696751096f917ee88f47aceb007

Funny thing is, that was actually your change :) So yah--I think the fix is to merge master into the branch, see those tests go red, and then change it to look for booleans.

Thanks for working with us on this! Let us know at any point if you need us to just jump in and take over, but until then thanks for making the changes for us and thanks for this PR!

**Comment by michaelrtm** [comment]
Thanks for pointing that out @mattstauffer! Getting better at git. Need to learn to not show so many commits here :)

I may have gone overboard with trailing commas on the arrays - added to single element arrays (should I have done that?).

Thanks for your support! Love that I can contribute :)

**Comment by imjohnbon** [comment]
@michaelrtm Yes, we also add trailing commas to arrays with single elements, so you're all good there. Thank you for all your hard work on this PR! We appreciate it.

**Pull request merged**